cli/src/pcluster/templates/cluster_stack.py

```diff
@@ -354,6 +385,113 @@ def _add_cleanup_resources_lambda(self):
+                        sid="SSMSendCommandTargetPolicy",
+                        effect=iam.Effect.ALLOW,
+                        actions=["ssm:SendCommand"],
+                        resources=[
+                            self.format_arn(
+                                service="ec2", account=self.account, region=self.region, resource="instance/*"
+                            ),
+                        ],
+                        # TODO Must restrict permission to target only the head node of the cluster.
+                        # The best approach would be tag-based, but the below conditions cause access denied errors.
+                        # conditions={
+                        #    "StringEquals": {
+                        #        "ec2:ResourceTag/parallelcluster:cluster-name": _cluster_name,
+                        #        "ec2:ResourceTag/parallelcluster:node-type": "HeadNode",
+                        #    },
+                        # }
+                    ),
+                    iam.PolicyStatement(
+                        sid="SSMGetCommandInvocationPolicy",
+                        effect=iam.Effect.ALLOW,
+                        actions=["ssm:GetCommandInvocation"],
+                        resources=[
+                            self.format_arn(service="ssm", account=self.account, region=self.region, resource="*"),
+                        ],
+                    ),
+                    get_cloud_watch_logs_policy_statement(
+                        resource=self.format_arn(service="logs", account="*", region="*", resource="*")
+                    ),
+                ]
+            ),
+            roles=[execution_role.ref],
+        )
+
+        self.cluster_admin_role.assume_role_policy_document.add_statements(
+            iam.PolicyStatement(
+                sid="TrustLambdaExecutionRolePolicy",
+                effect=iam.Effect.ALLOW,
+                actions=["sts:AssumeRole"],
+                principals=[iam.ArnPrincipal(execution_role.attr_arn)],
+            )
+        )
+
+        PclusterLambdaConstruct(
+            scope=self,
+            id=f"{function_id}FunctionConstruct",
+            function_id=function_id,
+            bucket=self.bucket,
+            config=self.config,
+            execution_role=execution_role.attr_arn,
+            handler_func="write_aws_credentials",
+            timeout=300,
+            memory_size=256,
+            environment={
+                "CLUSTER_NAME": _cluster_name,
+                "ROLE_ARN": self.cluster_admin_role.attr_arn,
+                "SSM_CW_LOG_GROUP_ENABLED": str(self.config.is_cw_logging_enabled),
+                "SSM_CW_LOG_GROUP_NAME": f"/aws/ssm/{_cluster_name}",
+            },
+            event_pattern={
+                "source": ["aws.ec2"],
+                "detail-type": ["EC2 Instance State-change Notification"],
+                "detail": {"state": ["running"], "instance-id": ["i-0fd6e84f8762a1675"]},
```

Comment:
should there be a ref to the head node rather than a hard-coded instance ID?

Reply:
I'm sorry: I forgot the hard wired testing value.